MusicNotationKit/MusicNotationKit/Measure.swift

```diff
@@ -53,50 +55,327 @@ public struct Measure: ImmutableMeasure, Equatable {
+        try newMeasure.insert(noteCollections, at: range.lowerBound, inSet: setIndex, skipTies: true)
+        self = newMeasure
+    }
+
+    public mutating func append(_ noteCollection: NoteCollection, inSet setIndex: Int = 0) {
+        notes[setIndex].append(noteCollection)
+    }
+
+    public mutating func insert(_ noteCollection: NoteCollection, at index: Int, inSet setIndex: Int = 0) throws {
+        try insert(noteCollection, at: index, inSet: setIndex, skipTies: false)
+    }
+
+    internal mutating func insert(_ noteCollections: [NoteCollection], at index: Int, inSet setIndex: Int = 0) throws {
+        try insert(noteCollections, at: index, inSet: setIndex, skipTies: false)
+    }
+
+    public mutating func removeNote(at index: Int, inSet setIndex: Int = 0) throws {
+        try removeNote(at: index, inSet: setIndex, skipTies: false)
+    }
 
+    public mutating func removeNotesInRange(_ indexRange: CountableClosedRange<Int>, inSet setIndex: Int = 0) throws {
+        try removeNotesInRange(indexRange, inSet: setIndex, skipTies: false)
+    }
+
+    /// Create a `Tuplet` from a note range. See `Tuplet.init` for more details.
+    /// This function also makes sure that the `noteRange` does not start or end
+    /// across a `Tuplet` boundary.
+    public mutating func createTuplet(_ count: Int, _ baseNoteDuration: NoteDuration, inSpaceOf baseCount: Int? = nil, fromNotesInRange noteRange: CountableClosedRange<Int>, inSet setIndex: Int = 0) throws {
+        var newMeasure = self
+        let startCollectionIndex = try newMeasure.noteCollectionIndex(fromNoteIndex: noteRange.lowerBound, inSet: setIndex)
+        if startCollectionIndex.tupletIndex != nil {
+            guard let tupletIndex = startCollectionIndex.tupletIndex, tupletIndex == 0 else {
+                throw MeasureError.invalidTupletIndex
+            }
+        }
+
+        let endCollectionIndex = try newMeasure.noteCollectionIndex(fromNoteIndex: noteRange.upperBound, inSet: setIndex)
+        if endCollectionIndex.tupletIndex != nil {
+            guard let tupletIndex = endCollectionIndex.tupletIndex,
+                let tuplet = newMeasure.notes[setIndex][endCollectionIndex.noteIndex] as? Tuplet else {
+                    assertionFailure("note collection should be tuplet, but cast failed")
+                    throw MeasureError.internalError
+            }
+            guard tuplet.noteCount == tupletIndex + 1 else {
+                throw MeasureError.invalidTupletIndex
+            }
+        }
+
+        var expectedCount = noteRange.upperBound - noteRange.lowerBound + 1
+        var tupletNotes = [NoteCollection]()
+        var index = startCollectionIndex.noteIndex
+
+        while expectedCount > 0 && notes[setIndex].count > index{
+            tupletNotes.append(notes[setIndex][index])
+            expectedCount -= notes[setIndex][index].noteCount
+            index += 1
+        }
```

Comment:
You should be able to create a range from `startCollectionIndex.noteIndex` and `endCollectionIndex.noteIndex` and then just do an array slice on `notes`. However, make sure you call `isValidIndexRange` first and throw an error if it's no good.